module/VuFind/src/VuFind/Auth/AbstractShibboleth.php

```diff
@@ -0,0 +1,372 @@
+     *
+     * @return array
+     */
+    abstract protected function getShibbolethConfiguration($request);
+
+    /**
+     * Extract required user attributes from the configuration.
+     *
+     * @param array $config shibboleth configuration
+     *
+     * @return array      Only username and attribute-related values
+     * @throws AuthException
+     */
+    protected function getRequiredAttributes($config)
+    {
+        // Special case -- store username as-is to establish return array:
+        $sortedUserAttributes = [];
+
+        // Now extract user attribute values:
+        foreach ($config as $key => $value) {
+            if (preg_match("/userattribute_[0-9]{1,}/", $key)) {
+                $valueKey = 'userattribute_value_' . substr($key, 14);
+                $sortedUserAttributes[$value] = $config[$valueKey] ?? null;
+
+                // Throw an exception if attributes are missing/empty.
+                if (empty($sortedUserAttributes[$value])) {
+                    throw new AuthException(
+                        "User attribute value of " . $value . " is missing!"
+                    );
+                }
+            }
+        }
+
+        return $sortedUserAttributes;
+    }
+
+    /**
+     * Add session id mapping to external_session table for single logout support
+     *
+     * @param \Laminas\Http\PhpEnvironment\Request $request Request object containing
+     * account credentials.
+     *
+     * @return \VuFind\Db\Row\User Object representing logged-in user.
+     */
+    protected function storeShibbolethSession($request)
+    {
+        $shib = $this->getConfig()->Shibboleth;
+        if (!isset($shib['session_id'])) {
+            return;
+        }
+        $shibSessionId = $this->getAttribute($request, $shib['session_id']);
+        if (null === $shibSessionId) {
+            return;
+        }
+        $localSessionId = $this->sessionManager->getId();
+        $externalSession = $this->getDbTableManager()
+            ->get('ExternalSession');
+        $externalSession->addSessionMapping(
+            $localSessionId, $shibSessionId
+        );
+        $this->debug(
+            "Cached Shibboleth session id '$shibSessionId' for local session"
+            . " '$localSessionId'"
+        );
+    }
+
+    /**
+     * Extract attribute from request.
+     *
+     * @param \Laminas\Http\PhpEnvironment\Request $request   Request object
+     * @param string                               $attribute Attribute name
+     *
+     * @throws AuthException
+     * @return string attribute value
+     */
+    protected function getAttribute($request, $attribute)
+    {
+        $proxy = $this->getConfig()->Shibboleth->proxy ?? false;
+        if ($proxy) {
+            $header = $request->getHeader($this->normalize($attribute));
+            return ($header) ? $header->getFieldValue() : null;
+        } else {
+            return $request->getServer()->get($attribute) ?? null;
```

Comment:
I don't think the `?? null` is needed on this line, since `get` will return null to begin with.